test/state/host.cpp

```diff
@@ -0,0 +1,369 @@
+
+    // Bump sender nonce.
+    if (msg.depth == 0 || msg.kind == EVMC_CREATE || msg.kind == EVMC_CREATE2)
+    {
+        if (sender_nonce == Account::NonceMax)
+            return {};  // Light early exception, cannot happen for depth == 0.
+        ++sender_acc.nonce;
+    }
+
+    if (msg.kind == EVMC_CREATE || msg.kind == EVMC_CREATE2)
+    {
+        // Compute and fill create address.
+        assert(msg.recipient == address{});
+        assert(msg.code_address == address{});
+        msg.recipient = compute_new_address(msg, sender_nonce);
+
+        // By EIP-2929, the  access to new created address is never reverted.
+        m_accessed_addresses.insert(msg.recipient);
+    }
+
+    return msg;
+}
+
+evmc::Result Host::create(const evmc_message& msg) noexcept
+{
+    assert(msg.kind == EVMC_CREATE || msg.kind == EVMC_CREATE2);
+
+    // Check collision as defined in pseudo-EIP https://github.com/ethereum/EIPs/issues/684.
+    // All combinations of conditions (nonce, code, storage) are tested.
+    if (const auto collision_acc = m_state.get_or_null(msg.recipient);
+        collision_acc != nullptr && (collision_acc->nonce != 0 || !collision_acc->code.empty()))
+        return evmc::Result{EVMC_OUT_OF_GAS};
+
+    auto& new_acc = m_state.get_or_create(msg.recipient);
+    if (m_rev >= EVMC_SPURIOUS_DRAGON)
+        new_acc.nonce = 1;
+    new_acc.storage.clear();  // In case of collision.
+
+    auto& sender_acc = m_state.get(msg.sender);  // TODO: Duplicated account lookup.
+    const auto value = intx::be::load<intx::uint256>(msg.value);
+    assert(sender_acc.balance >= value && "EVM must guarantee balance");
+    sender_acc.balance -= value;
+    new_acc.balance += value;  // The new account may be prefunded.
+
+    auto create_msg = msg;
+    create_msg.input_data = nullptr;
+    create_msg.input_size = 0;
+
+    // Execution can modify the state, iterators are invalidated.
+    auto result = m_vm.execute(*this, m_rev, create_msg, msg.input_data, msg.input_size);
+    if (result.status_code != EVMC_SUCCESS)
+    {
+        result.create_address = msg.recipient;
+        return result;
+    }
+
+    auto gas_left = result.gas_left;
+    assert(gas_left >= 0);
+
+    bytes_view code{result.output_data, result.output_size};
+    if (m_rev >= EVMC_SPURIOUS_DRAGON && code.size() > 0x6000)
+        return evmc::Result{EVMC_OUT_OF_GAS};
+
+    const auto cost = static_cast<int64_t>(code.size()) * 200;
+    gas_left -= cost;
+    if (gas_left < 0)
+    {
+        evmc::Result r{EVMC_OUT_OF_GAS};
+
+        if (m_rev == EVMC_FRONTIER)
+        {
+            r.status_code = EVMC_SUCCESS;
+            r.gas_left = result.gas_left;
+        }
+
+        return r;
+    }
+
+    // Reject EF code.
+    if (m_rev >= EVMC_LONDON && !code.empty() && code[0] == 0xEF)
+        return evmc::Result{EVMC_OUT_OF_GAS};
+
+    // TODO: The new_acc pointer is invalid because of the state revert implementation,
```

Comment:
Also not sure what this means

Reply:
They are invalidated when a call is reverted. We make a copy of unordered_map before the call and assign it back to the state in case of revert. This assign invalidates the pointers to accounts (they are replaced by the copy).

This is not happen if state journal is implemented. Then we keep the same unordered_map all the time.